src/Array.ts

```diff
@@ -736,13 +735,13 @@ export const rotate = <A>(n: number, xs: Array<A>): Array<A> => {
 }
 
 /**
- * Remove duplicate entries from an array, keeping the first found item.
+ * Remove duplicates from an array, keeping the first occurance of an element.
  * @function
  * @since 1.3.0
  */
-export const uniq = <A>(as: Array<A>, S: Setoid<A>): Array<A> => {
+export const uniq = <A>(S: Setoid<A>) => (as: Array<A>): Array<A> => {
   const r: Array<A> = []
-  fromArray(S)(as).forEach(a => r.push(a))
+  new Set(as).forEach(a => r.push(a))
```

Comment:
@alex-ketch that's not right, you're not using the Setoid in order to check for element equality.
If you try to use non primitive based setoid in your tests, you'll see it does not behave correctly.

on top of my head (not tested)

```typescript
import {
  insert,
  member
  } from 'fp-ts/lib/Set'
import { Setoid } from 'fp-ts/lib/Setoid'



export const unique = <T>(S: Setoid<T>) => {
  const setMember = member(S)
  const insertSet = insert(S)

  return (xs: T[]): T[] => {
    let set = new Set()
    const res = []
    for (const v of xs) {
      if (!setMember(set)(v)) {
        set = insertSet(v, set)
        res.push(v)
      }
    }
    return res
  }
}
```

Reply:
Also it may be very nice to return the original array when there was no duplicates.

Reply:
Oh.. I've given it more love, I think it answers the requirements.

```typescript
export const unique = <T>(S: Setoid<T>) => {
  const setMember = member(S)

  return (xs: T[]): T[] => {
    const set = new Set()
    const l = xs.length
    let i = 0
    for (; i < l; i++) {
      const v = xs[i]
      if (setMember(set)(v)) {
        break
      } else {
        set.add(v)
      }
    }
    if (i === l) {
      return xs
    } else {
      const r = xs.slice(0, i)
      for (; i < l; i++) {
        const v = xs[i]
        if (!setMember(set)(v)) {
          set.add(v)
          r.push(v)
        }
      }
      return r
    }
  }
}
```

tests:

```typescript
    interface Foo {
      a: string
      b: number
    }
    const fooSetoid = contramap((f: Foo) => f.b, ordNumber)
    const fooA: Foo = { a: 'a', b: 1 }
    const fooB: Foo = { a: 'b', b: 1 }
    const fooC: Foo = { a: 'c', b: 2 }
    const fooD: Foo = { a: 'd', b: 2 }

    const uniqueFooArr = [fooA, fooC]
    assert.equal(unique(fooSeoid)(uniqueFooArr), uniqueFooArr, 'preserve original array')
    assert.deepEqual(unique(fooSetoid)([fooA, fooB, fooC, fooD]), [fooA, fooC])
    assert.deepEqual(unique(fooSetoid)([fooB, fooA, fooC, fooD]), [fooB, fooC])
    assert.deepEqual(unique(fooSetoid)([fooA, fooA, fooC, fooD, fooA]), [fooA, fooC])
```

Reply:
Sorry @alex-ketch it may seems bully to push you such feedback like this..
I've been too tempted.

Reply:
All good @sledorze :) Thanks for pointing out the non-primitive use case.
I deviated a bit from your suggestion but let me know if you have any further feedback.

I know you mentioned that you didn't want a dependency on the `Set` module @gcanti, but now it's back with the `member` function.
I think it would be a bit too much if we duplicated all `member` dependencies. Is this ok?

Reply:
@alex-ketch we could add a `member` function for arrays

```ts
/**
 * Test if a value is a member of an array
 * @function
 * @since 1.3.0
 */
export const member = <A>(S: Setoid<A>) => (as: Array<A>, a: A): boolean => {
  return findIndex(as, e => S.equals(e, a)).isSome()
}

/**
 * Remove duplicates from an array, keeping the first occurance of an element.
 * @function
 * @since 1.3.0
 */
export const uniq = <A>(S: Setoid<A>): ((as: Array<A>) => Array<A>) => {
  const memberS = member(S)
  return as => {
    const r: Array<A> = []
    const len = as.length
    for (let i = 0; i < len; i++) {
      const a = as[i]
      if (!memberS(r, a)) {
        r.push(a)
      }
    }
    return r.length === len ? as : r
  }
}
```

Reply:
Would be nice to not have to allocate an option for each element..